### Carolina Garma Escoffié

## Counting characters function

In [1]:
def count_characters(string):
    diccionario = dict.fromkeys(string, 0)
    for char in string:
        diccionario[char] += 1 # access to the dict by key and sum 1
    return diccionario

In [2]:
count_characters('hola ! ? &, ')

{'h': 1, 'o': 1, 'l': 1, 'a': 1, ' ': 4, '!': 1, '?': 1, '&': 1, ',': 1}

## Instance

In [3]:
text="Empty spaces, what are we living for? Abandoned places, I guess we know the score, on and on Does anybody know what we are looking for? Another hero, another mindless crime Behind the curtain, in the pantomime Hold the line Does anybody want to take it anymore? The show must go on The show must go on, yeah Inside my heart is breaking My makeup may be flaking But my smile, still, stays on Whatever happens, I'll leave it all to chance Another heartache, another failed romance, on and on Does anybody know what we are living for? I guess I'm learning I must be warmer now I'll soon be turning, round the corner now Outside the dawn is breaking But inside in the dark I'm aching to be free The show must go on The show must go on Inside my heart is breaking My makeup may be flaking But my smile, still, stays on My soul is painted like the wings of butterflies Fairy tales of yesterday, grow but never die I can fly, my friends The show must go on The show must go on I'll face it with a grin I'm never giving in On with the show I'll top the bill I'll overkill I have to find the will to carry on On with the show Show Show must go on, go on, go on, go on, go on, go on, go on, go on"

In [4]:
text

"Empty spaces, what are we living for? Abandoned places, I guess we know the score, on and on Does anybody know what we are looking for? Another hero, another mindless crime Behind the curtain, in the pantomime Hold the line Does anybody want to take it anymore? The show must go on The show must go on, yeah Inside my heart is breaking My makeup may be flaking But my smile, still, stays on Whatever happens, I'll leave it all to chance Another heartache, another failed romance, on and on Does anybody know what we are living for? I guess I'm learning I must be warmer now I'll soon be turning, round the corner now Outside the dawn is breaking But inside in the dark I'm aching to be free The show must go on The show must go on Inside my heart is breaking My makeup may be flaking But my smile, still, stays on My soul is painted like the wings of butterflies Fairy tales of yesterday, grow but never die I can fly, my friends The show must go on The show must go on I'll face it with a grin I'm 

## Frequency list

In [5]:
frecuencia = count_characters(text)
{k: v for k, v in sorted(frecuencia.items(), key=lambda item: item[1])}

{'E': 1,
 'H': 1,
 'W': 1,
 'F': 1,
 'S': 2,
 'A': 3,
 'D': 3,
 'M': 3,
 'O': 3,
 '?': 4,
 'B': 4,
 'T': 6,
 "'": 8,
 'v': 9,
 'p': 10,
 'c': 14,
 'f': 14,
 'b': 15,
 'I': 15,
 'k': 15,
 'u': 22,
 ',': 23,
 'd': 23,
 'y': 26,
 'm': 30,
 'w': 31,
 'g': 32,
 'l': 41,
 'r': 43,
 'h': 50,
 's': 53,
 'i': 56,
 'a': 58,
 't': 63,
 'n': 81,
 'o': 87,
 'e': 96,
 ' ': 239}

## Probabilities

In [6]:
# Total number of characters
count=0
for key in frecuencia:
    count+=frecuencia[key]
print(count) 

1186


In [7]:
probabilities=dict.fromkeys(text,0)
for key in frecuencia:
    probabilities[key]=frecuencia[key]/count
probabilities # Show probabilities

{'E': 0.0008431703204047217,
 'm': 0.025295109612141653,
 'p': 0.008431703204047217,
 't': 0.05311973018549747,
 'y': 0.021922428330522766,
 ' ': 0.2015177065767285,
 's': 0.04468802698145025,
 'a': 0.048903878583473864,
 'c': 0.011804384485666104,
 'e': 0.08094435075885328,
 ',': 0.0193929173693086,
 'w': 0.026138279932546374,
 'h': 0.04215851602023609,
 'r': 0.03625632377740304,
 'l': 0.03456998313659359,
 'i': 0.047217537942664416,
 'v': 0.007588532883642495,
 'n': 0.06829679595278246,
 'g': 0.026981450252951095,
 'f': 0.011804384485666104,
 'o': 0.0733558178752108,
 '?': 0.003372681281618887,
 'A': 0.002529510961214165,
 'b': 0.012647554806070826,
 'd': 0.0193929173693086,
 'I': 0.012647554806070826,
 'u': 0.01854974704890388,
 'k': 0.012647554806070826,
 'D': 0.002529510961214165,
 'B': 0.003372681281618887,
 'H': 0.0008431703204047217,
 'T': 0.00505902192242833,
 'M': 0.002529510961214165,
 'W': 0.0008431703204047217,
 "'": 0.006745362563237774,
 'O': 0.002529510961214165,
 'F': 

In [8]:
total=0
for key in probabilities:
    total+=probabilities[key]
total # Sum must be 1

0.9999999999999999

## Huffman encoding

In [9]:
def encode(nodes, label, result, prefix = ''):  
    """
    Recursive function unction that creates the binary tree for the keys 
    OUTPUT EXAMPLE: {'0': {'0':'x', '1':'y'} , '1':'z'} 
    """
    childs = nodes[label]  # Acces to the childs labels of the given node
    
    tree = {} # Initialize binary tree as a dict
    
    if len(childs) == 2: # If the node has two children, start recurssion
        tree['0'] = encode(nodes, childs[0], result, prefix+'0') # left is 0
        tree['1'] = encode(nodes, childs[1], result, prefix+'1') # Right is 1 
        return tree # When recurssion ends, return the whole tree
    
    else:
        result[label] = prefix # If they are now leafs, assign the final prefix
        return label # Return the resultant label of the recurssion

In [10]:
def huffman(frequency_dict): # Receives an unsorted key frequency dictionary
    
    frequency_dict_copy = frequency_dict.copy() # Make a copy of the dict
    
    nodes = {} # nodes dict
    
    for label in frequency_dict_copy.keys(): # Leafs initialization, each label in keys-list will be a node
        
        nodes[label] = [] # Create a key-value node where key=label and value=[]

    while len(frequency_dict_copy) > 1: # binary tree creation
        
        frequency_dict_list = sorted(frequency_dict_copy.items(), key=lambda x:x[1]) # Create a sorted list by value (frequency)
        
        a1 = frequency_dict_list[0][0] # Smallest frequency label
        
        a2 = frequency_dict_list[1][0] # 2nd Smallest frequency label
        
        frequency_dict_copy[a1+a2] = frequency_dict_copy.pop(a1) + frequency_dict_copy.pop(a2) # Create a father with the sum of both popped children values
        
        nodes[a1+a2] = [a1, a2] # Create a node: label=join of both children, value=children original labels list
        
    code = {} # Initialize encoding
    
    root = a1+a2 # Final root with all the joined nodes
    
    tree = {} # Initialize tree
    
    tree = encode(nodes, root, code)   # recursive assignment of the code for the given binary tree    
    
    return code, tree # Return final enconding

## Encoding the song

In [11]:
code, tree = huffman(frecuencia)

In [12]:
code # clave

{' ': '00',
 'a': '0100',
 'c': '010100',
 'f': '010101',
 'm': '01011',
 'b': '011000',
 'I': '011001',
 'w': '01101',
 't': '0111',
 'k': '100000',
 "'": '1000010',
 'B': '10000110',
 'S': '100001110',
 'E': '1000011110',
 'H': '1000011111',
 'g': '10001',
 'n': '1001',
 'l': '10100',
 'v': '1010100',
 'p': '1010101',
 'u': '101011',
 'o': '1011',
 'r': '11000',
 ',': '110010',
 'd': '110011',
 'e': '1101',
 'h': '11100',
 'W': '1110100000',
 'F': '1110100001',
 'A': '111010001',
 'T': '11101001',
 'D': '111010100',
 'M': '111010101',
 'O': '111010110',
 '?': '111010111',
 'y': '111011',
 's': '11110',
 'i': '11111'}

In [13]:
tree

{'0': {'0': ' ',
  '1': {'0': {'0': 'a', '1': {'0': {'0': 'c', '1': 'f'}, '1': 'm'}},
   '1': {'0': {'0': {'0': 'b', '1': 'I'}, '1': 'w'}, '1': 't'}}},
 '1': {'0': {'0': {'0': {'0': {'0': 'k',
      '1': {'0': "'",
       '1': {'0': 'B', '1': {'0': 'S', '1': {'0': 'E', '1': 'H'}}}}},
     '1': 'g'},
    '1': 'n'},
   '1': {'0': {'0': 'l', '1': {'0': {'0': 'v', '1': 'p'}, '1': 'u'}},
    '1': 'o'}},
  '1': {'0': {'0': {'0': 'r', '1': {'0': ',', '1': 'd'}}, '1': 'e'},
   '1': {'0': {'0': 'h',
     '1': {'0': {'0': {'0': {'0': {'0': 'W', '1': 'F'}, '1': 'A'}, '1': 'T'},
       '1': {'0': {'0': 'D', '1': 'M'}, '1': {'0': 'O', '1': '?'}}},
      '1': 'y'}},
    '1': {'0': 's', '1': 'i'}}}}}

## Encoded file

In [14]:
def encode_text(text,code):
    encoded = ''.join([code[t] for t in text])
    return(encoded)

In [15]:
encoded_text=encode_text(text,code)
print('Encoded text:')
print(encoded_text)

Encoded text:
10000111100101110101010111111011001111010101010100010100110111110110010000110111100010001110001001100011010001101110100101001111110101001111110011000100010101101111000111010111001110100010110000100100111001110111001110111001100101010110100010001010011011111011001000011001001000110101111011111011110000110111010010000010011011011010001111110011010011110010100101111000110111001000101110010001001001110011001011100100111010100101111011111000010010011110110110001011110011111011001000001001101101101000110111100010001110001101110100010011000110100101001011101110000011111100110001000101011011110001110101110011101000110011011011111100110111000001110011011100010111100100001001001101101111110011011100000010111111110011100111010011011111011110000101001100011111010111101001000011011011110011111100111001100011111100110100010100101011110000111010011111100111001000111111001000111111001101001010101010010010111101101011111110101111010010000111111011101001100110001111110011010010100111111001

## Decoding the song

In [16]:
def decoding(encoded_text, tree):
    
    decoded = [] # Empty decoded string list
    i = 0 # index counter
    
    while i < len(encoded_text): # decoding using the binary tree search
        children = encoded_text[i]  
        level = tree[children] # Accessing to the levels
        
        while (type(level)!=str): # While level != string, do DST
            i += 1 
            children = encoded_text[i] # Accessing to the next level
            level = level[children] 
            
        decoded.append(level) # Add the decoded string to the list     
        i += 1 # Start with the next binary sequence
    
    return ''.join(decoded)

In [17]:
decoded=decoding(encoded_text, tree)
print('Decoded text:')
print(decoded)

Decoded text:
Empty spaces, what are we living for? Abandoned places, I guess we know the score, on and on Does anybody know what we are looking for? Another hero, another mindless crime Behind the curtain, in the pantomime Hold the line Does anybody want to take it anymore? The show must go on The show must go on, yeah Inside my heart is breaking My makeup may be flaking But my smile, still, stays on Whatever happens, I'll leave it all to chance Another heartache, another failed romance, on and on Does anybody know what we are living for? I guess I'm learning I must be warmer now I'll soon be turning, round the corner now Outside the dawn is breaking But inside in the dark I'm aching to be free The show must go on The show must go on Inside my heart is breaking My makeup may be flaking But my smile, still, stays on My soul is painted like the wings of butterflies Fairy tales of yesterday, grow but never die I can fly, my friends The show must go on The show must go on I'll face it wit

## Number of bits

#### Normal encoding

In [18]:
len(text)*8 # Bites in normal encoding

9488

#### Huffman encoding

In [19]:
def count_bin_bits(frecuencia, clave):
    count=0 # bits counter
    for key in frecuencia:
        count+=frecuencia[key]*len(clave[key])
    return count

In [20]:
count_bin_bits(frecuencia, code)

5171

## By-hand enconding

Using my by-hand encoding, I also got 5171 bits for the encoded song. Let's see if my "clave" dict also encodes and decodes well..

<img src="my_binary_tree.jpg">

In [21]:
my_tree= {'0': {'0': ' ',
                '1': {'0': {'0': {'0': {'0': 'c', '1': 'f'}, '1': 'm'}, '1': 'a'},
                      '1': {'0': {'0': {'0': 'I', '1': 'b'}, '1': 'w'}, '1':'t'}}},
          '1':  {'0': {'0': {'0': {'0': {'0': 'k', '1': {'0': {'0': 'B', '1': {'0': 'S', '1': {'0': 'E', '1': 'H'}}}, '1': "'"}}, '1': 'g'}, '1': 'n'},
                       '1': {'0': {'0': 'l', '1': {'0': {'0': 'v', '1': 'p'}, '1': 'u'}}, '1': 'o'}},
                 '1': {'0': {'0': {'0': 'r', '1': {'0': ',', '1': 'd'}}, '1': 'e'},
                       '1':  {'0': {'0': 'h', '1': {'0': {'0': {'0': 'T', '1': {'0': {'0': 'W', '1': 'F'}, '1':'A'}},
                                                          '1': {'0': {'0': 'D', '1': 'M'},
                                                                '1': {'0': 'O', '1': '?'}}},
                                                   '1': 'y'}},
                              '1': {'0': 's', '1': 'i'}}}}}

In [22]:
mi_clave = {
    'E':'1000010110',
    'H':'1000010111',
    'W':'1110100100',
    'F':'1110100101',
    'S':'100001010',
    'A':'111010011',
    'D':'111010100',
    'M':'111010101',
    'O':'111010110',
    '?':'111010111',
    'B':'10000100',
    'T':'11101000',
    "'":'1000011',
    'v':'1010100',
    'p':'1010101',
    'c':'010000',
    'f':'010001',
    'b':'011001',
    'I':'011000',
    'k':'100000',
    'u':'101011',
    ',':'110010',
    'd':'110011',
    'y':'111011',
    'm':'01001',
    'w':'01101',
    'g':'10001',
    'l':'10100',
    'r':'11000',
    'h':'11100',
    's':'11110',
    'i':'11111',
    'a':'0101',
    't':'0111',
    'n':'1001',
    'o':'1011',
    'e':'1101',
    ' ':'00'   
}

In [23]:
count_bin_bits(frecuencia, mi_clave)

5171

In [24]:
encoded_text = encode_text(text, mi_clave)
encoded_text

'100001011001001101010101111110110011110101010101010100001101111101100100001101111000101011100010111000110100011011101001010011111101010011111100110001000100011011110001110101110011101001101100101011001110011101110011101110011001010101101000101010000110111110110010000110000010001101011110111110111100001101110100100000100110110110100011111100110100111100100001011110001101110010001011100100010110011100110010111001001110101001011110111110000101100111101101100110111100111110110010000010011011011010001101111000101011100011011101000101110001101001010010111011100000111111001100010001000110111100011101011100111010011100110110111111001101110000011100110111000101111001000010110011011011111100110111000000100111111100111001110100110111110111100001000011000111110100111010010000100110111100111111001110011000111111001101000100001010111100001110101111111001110010001111110010001111110011010010101010101100101111011010011111101001110100100001011110111010011001100011111100110100101001111110011101001110101

In [25]:
decoding(encoded_text, my_tree)

"Empty spaces, what are we living for? Abandoned places, I guess we know the score, on and on Does anybody know what we are looking for? Another hero, another mindless crime Behind the curtain, in the pantomime Hold the line Does anybody want to take it anymore? The show must go on The show must go on, yeah Inside my heart is breaking My makeup may be flaking But my smile, still, stays on Whatever happens, I'll leave it all to chance Another heartache, another failed romance, on and on Does anybody know what we are living for? I guess I'm learning I must be warmer now I'll soon be turning, round the corner now Outside the dawn is breaking But inside in the dark I'm aching to be free The show must go on The show must go on Inside my heart is breaking My makeup may be flaking But my smile, still, stays on My soul is painted like the wings of butterflies Fairy tales of yesterday, grow but never die I can fly, my friends The show must go on The show must go on I'll face it with a grin I'm 

## Check if my by-hand decoding is equal to the script's decoded string

In [26]:
decoding(encode_text(text,code), tree)==decoding(encode_text(text,mi_clave), my_tree)

True

In [27]:
text==decoding(encode_text(text,code), tree) 

True

In [28]:
text==decoding(encode_text(text,mi_clave), my_tree)

True

## That means that both encoding methods, script and by-hand, were a success!